source/common/network/win32_socket_handle_impl.cc

```diff
@@ -71,10 +79,39 @@ Api::IoCallUint64Result Win32SocketHandleImpl::recvmmsg(RawSliceArrays& slices,
 }
 
 Api::IoCallUint64Result Win32SocketHandleImpl::recv(void* buffer, size_t length, int flags) {
-
-  Api::IoCallUint64Result result = IoSocketHandleImpl::recv(buffer, length, flags);
-  reEnableEventBasedOnIOResult(result, Event::FileReadyType::Read);
-  return result;
+  if (flags & MSG_PEEK) {
+    // can a remote OOM us now that we are not protected by readDisable?
+    Api::IoCallUint64Result peek_result = drainToPeekBuffer();
+    //  Some fatal error happened
+    if (!peek_result.wouldBlock()) {
+      return peek_result;
+    }
+
+    // No data available, register read again.
+    if (peek_result.wouldBlock() && peek_buffer_->length() == 0) {
+      if (file_event_) {
+        file_event_->registerEventIfEmulatedEdge(Event::FileReadyType::Read);
+      }
+      return peek_result;
+    }
+
+    Api::IoCallUint64Result result = peekFromPeekBuffer(buffer, length);
+    if (peek_buffer_->length() < length && file_event_) {
+      file_event_->registerEventIfEmulatedEdge(Event::FileReadyType::Read);
+    } else {
+      // This means that our peak buffer has more data than what the user
+      // wanted. Return the slice to the caller.
+      // How can the caller (v2 proxy protocol inspector) reactivate the events again here?
```

Comment:
If the listener filter got enough data, it shouldn't reactivate the events again.

Reply:
I am wondering if there is a listener that wants to do the following logic (I am not sure if such a listener exists, but it seems valid to me).

1. Listener asks to peek 10 bytes
2. io_handle reads from the wire 25 bytes
3.Listener processes these 10 bytes and realize that it needs to peek another 50 bytes

Who is responsible to reactivate the events in this case?

Reply:
Probably we need the listener filter to trigger that. But we don't have any listener filter doing the case you said today. Currently all the listener filter has max bytes they want to peek. I also define an interface for listener filter to define the max bytes they want https://github.com/envoyproxy/envoy/pull/17395/files#diff-84c1ba56d718c40a4f8cb0204c28b8a627a9d644775c80af72a37b5d62ef70f2R90, so it guarantees there won't have such case.